avocado/core/runners/requirement_asset.py

```diff
@@ -0,0 +1,108 @@
+
+     * uri: not used
+
+     * args: not used
+
+     * kwargs:
+        - name: the file name or uri (required)
+        - asset_hash: hash of the file (optional)
+        - algorithm: hash algorithm (optional)
+        - locations: location(s) where the file can be fetched from (optional)
+        - expire: time in seconds for the asset to expire (optional)
+    """
+
+    @staticmethod
+    def _fetch_asset(name, asset_hash, algorithm, locations, expire, queue):
+        cache_dirs = settings.as_dict().get('datadir.paths.cache_dirs')
+
+        asset_manager = Asset(name, asset_hash, algorithm, locations,
+                              cache_dirs, expire)
+
+        result = 'pass'
+        stdout = ''
+        stderr = ''
+        try:
+            asset_file = asset_manager.fetch()
+            stdout = 'File fetched at %s' % asset_file
+        except OSError as exc:
+            result = 'error'
+            stderr = str(exc)
+
+        output = {'result': result,
+                  'stdout': stdout,
+                  'stderr': stderr}
+        queue.put(output)
+
+    def run(self):
+        yield self.prepare_status('started')
+
+        name = self.runnable.kwargs.get('name')
+        asset_hash = self.runnable.kwargs.get('asset_hash')
+        algorithm = self.runnable.kwargs.get('algorithm')
+        locations = self.runnable.kwargs.get('locations')
+        expire = self.runnable.kwargs.get('expire')
+        if expire is not None:
+            expire = data_structures.time_to_seconds(str(expire))
+        # if name was passed correctly, run the Avocado Asset utility
+        if name is not None:
+            # let's spawn it to another process to be able to update the
+            # status messages and avoid the Asset to lock this process
+            queue = SimpleQueue()
+            process = Process(target=self._fetch_asset,
+                              args=(name, asset_hash, algorithm, locations,
+                                    expire, queue))
+            process.start()
+
+            while queue.empty():
+                time.sleep(nrunner.RUNNER_RUN_STATUS_INTERVAL)
+                yield self.prepare_status('running')
+
+            output = queue.get()
+            result = output['result']
+            stdout = output['stdout']
+            stderr = output['stderr']
+        else:
+            # Otherwise, log the missing package name
+            result = 'error'
+            stdout = ''
+            stderr = ('At least name should be passed as kwargs using'
+                      ' name="uri".')
+
+        yield self.prepare_status('running',
+                                  {'type': 'stdout',
+                                   'log': stdout.encode()})
+        yield self.prepare_status('running',
+                                  {'type': 'stderr',
+                                   'log': stderr.encode()})
+        yield self.prepare_status('finished', {'result': result})
```

Comment:
This sounds a little bit odd to me.

Should we have a "running" even when there is no name? I have the impression that the name is None check could be one of the first things.

```python
if name is None:
   return self.prepare_statys('finished', {'result': 'cancel', 'fail_reason': 'At least name should be passed as kwargs using ...'})
```

Also, we have on other runners the '"fail_reason" that AFAICT would be to separate a normal error from a nrunner code validation. I would recommend to leave stderr type during running only when the subprocess fails. Feels to me that "abort/cancel" would be the most appropriate error here, if we haven't started the "fetch_asset()" method yet.

Reply:
Right now we don't have `fail_reson` defined for runners. We currently have: https://avocado-framework.readthedocs.io/en/latest/guides/contributor/chapters/runners.html#supported-message-types

If the runner fails, independent of where it failed, I prefer to set the task as failed instead of anything else to avoid confusion. Runner status is related to the task itself, not to a test. abort/cancel are related to a test run.

Reply:
I mean, this runner fail is related to a non-test task.

Reply:
> This sounds a little bit odd to me.
> 
> Should we have a "running" even when there is no name? I have the impression that the name is None check could be one of the first things.

I forgot to comment, there won't be a `running` status when `name` is `None`:

```
$ avocado-runner-requirement-asset runnable-run
{'status': 'started', 'time': 57914.857953882}
{'type': 'stdout', 'log': b'', 'status': 'running', 'time': 57914.857973143}
{'type': 'stderr', 'log': b'At least name should be passed as kwargs using name="uri".', 'status': 'running', 'time': 57914.857979178}
{'result': 'error', 'status': 'finished', 'time': 57914.857983989}
```

Reply:
> Right now we don't have `fail_reson` defined for runners. We currently have: https://avocado-framework.readthedocs.io/en/latest/guides/contributor/chapters/runners.html#supported-message-types
> 

It looks like we have wrong documentation, because the `fail_reson` is working properly for runners and for example the `DryRunRunner` is using that https://github.com/avocado-framework/avocado/blob/master/avocado/core/nrunner.py#L388

Reply:
So, I was thinking again about it and I recalled why I did it the way it is now. 

The requirements resolver will create the non-test tasks to fulfill requirements. Remember we had to change how we handle messages to avoid non-tests to do not be accounted as tests here: https://github.com/avocado-framework/avocado/pull/4546/commits/c8613cb90d321a36ddfa2f11fa4dd341714d3021

A requirement runner is not a utility, its objective is to fulfill a requirement. For a utility execution, I agree we should have more than a success/fail output, but for fulfillment of a requirement, what matters for the state machine is a pass/not pass state as shown here https://github.com/avocado-framework/avocado/blob/c8338d8c729b07f64287bd4c5a8bb3e064836891/avocado/core/task/statemachine.py#L101

If a requirement task fails, the only place we will have logs is the `test-results` directory. I don't feel right adding anything other than success messages to the `stdout` file. I would not look for error messages there. Any message for the failed status will go to the only remaining file we have for logs, the `stderr` file.

This makes things simple for requirements runners. Again, if we were talking about a utility, then I agree we should have a more elaborated output with different error handling.